File: data/shared/lab04/work/gdd-spark/scala/AverageSeriesRating.scala
import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession

// in spark shell, sc is defined for you
// val sc = new SparkContext(new SparkConf().setAppName("Scalalab5"))

// you can change this to run on the full data
val filein = "hdfs://cm:9000/uhadoop/shared/imdb/imdb-ratings-two.tsv"
val fileout = "hdfs://cm:9000/uhadoop2024/<user>/series-avg-two-sc/"

val input = sc.textFile(filein)
// to dump an RDD like input to the shell, use
//   input.collect()
// make sure not to do this for a large RDD
    
val lines = input.map(line => line.split("\t"))

val tvSeries = lines.filter(line => line(4) == "tvSeries" && line(5) != "null")

var seriesEpisodeRating = tvSeries.map(line => (line(2) + "#" + line(3), line(5), line(1).toFloat))

var seriesToEpisodeRating = seriesEpisodeRating.map(tup => (tup._1, tup._3))

var seriesToSumCountRating = seriesToEpisodeRating.aggregateByKey((0f, 0)) ((sumCount: (Float,Int), rating: Float) => (sumCount._1 + rating, sumCount._2 + 1), (sumCountA: (Float,Int), sumCountB: (Float,Int)) => (sumCountA._1 + sumCountB._1, sumCountA._2 + sumCountB._2) )

var seriesToAvgRating = seriesToSumCountRating.map(tup => (tup._1,tup._2._1/tup._2._2) )

seriesToAvgRating.saveAsTextFile(fileout)

sc.stop()

// to quit the shell:
//   :quit

File: data/shared/lab04/work/gdd-spark/py/average_series_rating.py
from __future__ import print_function

import sys
from pyspark.sql import SparkSession

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: average_series_rating.py <filein> <fileout>", file=sys.stderr)
        sys.exit(-1)

    filein = sys.argv[1]
    fileout = sys.argv[2]
    
    # in pyspark shell start with:
    #   filein = "hdfs://cm:9000/uhadoop/shared/imdb/imdb-ratings-two.tsv"
    #   fileout = "hdfs://cm:9000/uhadoop2025/<user>/series-avg-two-py/"
    # and continue line-by-line from here

    spark = SparkSession.builder.appName("AverageSeriesRating").getOrCreate()

    input = spark.read.text(filein).rdd.map(lambda r: r[0])
    # to dump an RDD like input to the pyspark shell, use
    #   input.collect()
    # make sure not to do this for a large RDD

    lines = input.map(lambda line: line.split("\t"))

    tvSeries = lines.filter(lambda line: ("tvSeries" == line[4]) and not ('null' == line[5]))

    seriesEpisodeRating = tvSeries.map(lambda line: (line[2]+ "#" + line[3], line[5], float(line[1])))

    seriesToEpisodeRating = seriesEpisodeRating.map(lambda tup: (tup[0], tup[2]))

    seriesToSumCountRating = seriesToEpisodeRating.aggregateByKey((0.0, 0), \
        lambda sumCount, rating: (sumCount[0] + rating, sumCount[1] + 1), \
        lambda sumCountA, sumCountB: (sumCountA[0] + sumCountB[0], sumCountA[1] + sumCountB[1]))

    seriesToAvgRating = seriesToSumCountRating.mapValues(lambda tup2n: tup2n[0]/tup2n[1])

    seriesToAvgRating.saveAsTextFile(fileout)

    spark.stop()